packages/devextreme/js/__internal/ui/popup/m_popup.ts

```diff
@@ -432,121 +467,244 @@ class Popup<
   }
 
-  _renderTemplateByType(optionName, data, $container, additionalToolbarOptions?): dxElementWrapper {
-    const {
-      rtlEnabled, useDefaultToolbarButtons, useFlatToolbarButtons, disabled,
-    } = this.option();
-    const template = this._getTemplateByOption(optionName);
-    const toolbarTemplate = template instanceof EmptyTemplate;
-
-    if (toolbarTemplate) {
-      const integrationOptions = extend({}, this.option('integrationOptions'), { skipTemplates: ['content', 'title'] });
-      const toolbarOptions = extend(additionalToolbarOptions, {
-        items: data,
-        rtlEnabled,
-        useDefaultButtons: useDefaultToolbarButtons,
-        useFlatButtons: useFlatToolbarButtons,
-        disabled,
-        integrationOptions,
-      });
+  _renderBottomToolbar(): void {
+    const items = this._getToolbarItems('bottom');
+
+    if (!items.length) {
+      this._$bottomToolbar?.detach();
+      return;
+    }
 
-      this._getTemplate('dx-polymorph-widget').render({
-        container: $container,
-        model: {
-          widget: this._getToolbarName(),
-          options: toolbarOptions,
+    if (this._$bottomToolbar) {
+      this._updateToolbarOptions('bottom', { items });
+    } else {
+      this._renderBottomToolbarImpl();
+    }
+    // To trigger toolbar width set in initial rendering to set menu button width (T1245421)
+    // And to trigger with animation in runtime items update
+    this._triggerToolbarResizeEvent();
+    this._toggleClasses();
+  }
+
+  _renderBottomToolbarImpl(): void {
+    this._$bottomToolbar?.remove();
+
+    const items = this._getToolbarItems('bottom');
+    const $toolbarContainer = $('<div>')
+      .addClass(POPUP_BOTTOM_CLASS)
+      .addClass(TOOLBAR_CLASS)
+      .insertAfter(this.$content());
+
+    this._$bottomToolbar = this._renderToolbar(
+      'bottomTemplate',
+      items,
+      $toolbarContainer,
+      {
+        compactMode: true,
+        onInitialized: (e): void => {
+          this._bottomToolbar = e.component;
         },
-      });
-      const $toolbar = $container.children('div');
-      $container.replaceWith($toolbar);
-      return $toolbar;
+      },
+    );
+
+    this._$bottomToolbar.addClass(POPUP_BOTTOM_CLASS);
+  }
+
+  _triggerToolbarResizeEvent(): void {
+    triggerResizeEvent(this.$overlayContent());
+    triggerResizeEvent(this.$overlayContent());
+  }
+
+  _renderToolbar(
+    optionName: string,
+    items: ToolbarItem[],
+    $container: dxElementWrapper,
+    additionalToolbarOptions: Partial<ToolbarBaseProperties> = {},
+  ): dxElementWrapper {
+    const template = this._getTemplateByOption(optionName);
+    const isEmptyTemplate = template instanceof EmptyTemplate;
+
+    if (isEmptyTemplate) {
+      return this._renderByPolymorphTemplate(items, $container, additionalToolbarOptions);
     }
+
+    return this._renderByTemplate(template, $container);
+  }
+
+  _renderByPolymorphTemplate(
```

Comment:
I believe it should be (but I see that it's already called this way and there's a CSS class, so we can leave as is)
```suggestion
  _renderByPolymorphicTemplate(
```

Reply:
Improved here: [f12314916353b33f10d4c7ee4c99a008bcb9d6a4]